File: Python/601-650/TwoKeysKeyboard.py
class Solution(object):
    def minSteps(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [ i for i in range(1, n + 1) ]
        dp[0] = 0
        for i in range(1, n // 2 + 1):
            for j in range(i * 2, n + 1, i):
                dp[j - 1] = min(dp[j - 1], dp[i - 1] + j / i)
        return dp[n - 1]

File: Java/601-650/TwoKeysKeyboard.java
class TwoKeysKeyboard {
    public int minSteps(int n) {
        int[] dp = new int[n + 1];
        dp[1] = 0;
        for (int i = 1; i <= n / 2; i++) {
            for (int j = i + i; j <= n; j += i) {
                dp[j] = dp[i] + j / i;
            }
        }
        return dp[n];
    }
}
